.github/workflows/builder_docker-based_slsa3.yml

```diff
@@ -447,11 +469,63 @@ jobs:
       - name: Upload the signed attestations
         id: upload
         if: ${{ github.event_name != 'pull_request' }}
-        uses: actions/upload-artifact@0b7f8abb1508181956e8e162db84b466c27e18ce # v3.1.2
+        uses: ./__BUILDER_CHECKOUT_DIR__/.github/actions/secure-upload-artifact
         with:
           name: "${{ env.OUTPUT_FOLDER }}-${{ needs.rng.outputs.value }}"
           path: "${{ env.OUTPUT_FOLDER }}-${{ needs.rng.outputs.value }}"
-          if-no-files-found: error
+
+  ###################################################################
+  #                                                                 #
+  #          Upload artifacts and provenance as assets              #
+  #                                                                 #
+  ###################################################################
+  upload-assets:
+    permissions:
+      contents: write # Needed to write artifacts to a release.
+    runs-on: ubuntu-latest
+    needs: [build, provenance, detect-env]
+    if: inputs.upload-assets && (startsWith(github.ref, 'refs/tags/') || inputs.upload-tag-name != '')
+    steps:
+      - name: Checkout builder repository
+        uses: slsa-framework/slsa-github-generator/.github/actions/secure-builder-checkout@main
+        with:
+          repository: "${{ needs.detect-env.outputs.repository }}"
+          ref: "${{ needs.detect-env.outputs.ref }}"
+          path: __BUILDER_CHECKOUT_DIR__
+
+      # TODO(https://github.com/slsa-framework/slsa-github-generator/issues/1655): Investigate
+      # use of secure download for a folder of artifacts, or use the SLSA layout files and their
+      # checksums.
+      - name: Download artifacts
+        uses: actions/download-artifact@9bc31d5ccc31df68ecc42ccf4149144866c47d8a # v3.0.2
+        with:
+          name: "${{ needs.build.outputs.build-outputs-name }}"
+          path: "${{ needs.build.outputs.build-outputs-name }}"
+
+      - name: Download provenance
+        uses: ./__BUILDER_CHECKOUT_DIR__/.github/actions/secure-download-artifact
+        with:
+          name: "${{ needs.provenance.outputs.provenance-name }}"
+          path: "${{ needs.provenance.outputs.provenance-name }}"
+          sha256: "${{ needs.provenance.outputs.provenance-sha256 }}"
+
+      - name: Upload provenance new tag
+        uses: softprops/action-gh-release@de2c0eb89ae2a093876385947365aca7b0e5f844 # v0.1.15
+        if: startsWith(github.ref, 'refs/tags/') && inputs.upload-tag-name == ''
+        id: release-new-tags
+        with:
+          prerelease: ${{ inputs.prerelease }}
+          files: |
+            ${{ needs.build-dry.outputs.go-binary-name }}
```

Comment:
This looks like a CnP from the go builder? maybe this should be `${{ needs.build.outputs.build-outputs-name }}` ?